impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -600,10 +598,28 @@ public float getHitRate(TimeUnit unit) {
           // todo: not hinting enforceCapacity() about the mapping we just added makes it likely that it will be the eviction target
           enforceCapacity(1);
           try {
-            ValueHolder<V> value = fault.get();
+            final ValueHolder<V> value = fault.get();
             final OnHeapValueHolder<V> newValue;
             if(value != null) {
-              newValue = importValueFromLowerTier(key, value, now);
+              try {
+                newValue = importValueFromLowerTier(key, value, now);
+              } catch (RuntimeException re) {
+                LOG.error("Expiry computation caused an exception - Expiry duration will be 0 ", re);
+                map.computeIfPresent(key, new BiFunction<K, OnHeapValueHolder<V>, OnHeapValueHolder<V>>() {
+                  @Override
+                  public OnHeapValueHolder<V> apply(K mappedKey, final OnHeapValueHolder<V> mappedValue) {
+                    if(mappedValue.equals(fault)) {
+                      final AbstractValueHolder<V> valueHolder = (AbstractValueHolder<V>)value;
```

Comment:
I do not like the blind cast. I suggest doing something like:
 - check if expired already - if yes, call into onExpiration directly
 - if not check if `instanceof AbstractValueHolder` and if yes update expiration time
 - in case it is not the type we expect, do nothing as we have no way of sending expiry info to authority